text/3417-dropck-eyepatch-v3.md

```diff
@@ -0,0 +1,239 @@
+- Feature Name: (`dropck_eyepatch_v3`)
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Cleanup the rules for implicit drops by adding an argument for `#[may_dangle]` on type
+parameters: `#[may_dangle(droppable)]` and `#[may_dangle(must_not_use)]`. Change `PhantomData`
+to get completely ignored by dropck as its current behavior is confusing and inconsistent.
+
+# Motivation
+[motivation]: #motivation
+
+The current rules around dropck and `#[may_dangle]` are confusing and have even resulted in
+unsoundness [multiple](https://github.com/rust-lang/rust/issues/76367)
+[times](https://github.com/rust-lang/rust/issues/99408). Even without `#[may_dangle]`,
```

Comment:
Re-phrasing the key point of this RFC at a slightly higher level, the argument seems to be: inferring the exact effect of `may_dangle` through field ownership is subtle and easy to get wrong, we should instead explicitly annotate what we want.

I generally agree with this direction. However, this is a departure from what seems to me as a pretty fundamental design decision in Rust: to infer all sorts of properties of a type based on "which other types it owns". We do this for dropck, but also for auto traits and variance. This decisions leads to issues such as https://github.com/rust-lang/rust/issues/99408 where changing the type of a field accidentally also changes this inferred property (that issue was `T` to `ManuallyDrop<T>` affecting dropck; I think we also had `&mut T` to `*mut T` affecting auto traits -- lucky enough erring on the safe side -- and it's easy to imagine similar situations with variance).

I am not sure how I feel about attacking this fundamental problem in a piecemeal fashion. At the very least, this RFC should state somewhere that the question of inferring may_dangle dropck from field ownership is closely related to inferring auto traits and variance. And long-term, IMO it would be very odd to have this explicit annotation for may_dangle without also having something similar for auto traits and variance. Admittedly that's a lot harder, (a) since those are stable and (b) since we only want this more explicit style when unsafe code is involved -- which for may_dangle is 100% of the cases, but not so for auto traits and variance.

Put differently: if PhantomData is not fit for the purpose of may_dangle, why is it ever fit for any purpose? Given the arguments in this RFC, it seems wrong to keep that type around (well it will stay for backwards compatibility obviously). IMO we should only accept the RFC if we generally agree that PhantomData should be deprecated and replaced by more explicit annotations.

Reply:
we feel like there's no room for error with unsafe. inference isn't inherently wrong, but it should fail-safe if you forget something. that appears to hold for `Send` and `Sync` but not so much for `may_dangle`.

Reply:
> that appears to hold for Send and Sync

I don't think so. If you have a `*mut T` field and later decide to make it `&mut`, you might be accidentally adding `Send`/`Sync` to your type without realizing. Similar examples exist for variance.

Reply:
that's not an example, that's an argument.

we can't think of an example where changing `*mut T` to `&mut T` causes it to become unsound due to `Send`/`Sync`, but plenty of examples where it becomes unsound due to aliasing. we would love to see such an example tho.

on the other hand variance *is* a footgun, from experience. at the same time, variance is fairly unintuitive, extremely verbose, and it shows up *everywhere*, so there's a huge added value from having it inferred by default. this is in contrast to `#[may_dangle]` which (on stable) only shows up on std collections and pointer wrappers.

further, `#[may_dangle]` opts-*in* to inference (`impl Drop` is the opt-out), whereas there's no way to opt-*out* of variance.

Reply:
> we can't think of an example where changing `*mut T` to `&mut T` causes it to become unsound due to `Send`/`Sync`, but plenty of examples where it becomes unsound due to aliasing. we would love to see such an example tho.

There is such an example: https://github.com/rust-lang/rust/issues/41622

> Put differently: if PhantomData is not fit for the purpose of may_dangle, why is it ever fit for any purpose? Given the arguments in this RFC, it seems wrong to keep that type around (well it will stay for backwards compatibility obviously). IMO we should only accept the RFC if we generally agree that PhantomData should be deprecated and replaced by more explicit annotations.

Going to think more about that and will try to come back with an answer later.

Reply:
> IMO we should only accept the RFC if we generally agree that PhantomData should be deprecated and replaced by more explicit annotations.

I worded this a bit strongly; a point can probably be made for this RFC while still keeping PhantomData for other purposes. But the RFC should then make that point, since the symptoms it describes apply more generally to field-ownership-based inference, not just to may_dangle / dropck.

Reply:
> There is such an example: [rust-lang/rust#41622](https://github.com/rust-lang/rust/issues/41622)

Hm? We don't see it. That issue doesn't have any backtracking of how/when it regressed.

Reply:
It wasn't a regression, but it was a case of accidental `Sync` inference: `MutexGuard` used `&mut` which propagates `Sync`; if it had used `*mut` (which arguably would be more correct as well since the lifetime of the reference is not accurate) the bug would have been avoided.

Reply:
Hmm, so maybe we should have more opt-outs, but having an opt-out mixed with a partial opt-in seems like a massive footgun, especially if it's not actually checked for correctness.

But also if custom auto traits are ever to be a thing, how do they fit into all this?

Reply:
Alright, here are my thoughts after thinking about this during lunch:

> Put differently: if PhantomData is not fit for the purpose of may_dangle, why is it ever fit for any purpose? Given the arguments in this RFC, it seems wrong to keep that type around (well it will stay for backwards compatibility obviously). IMO we should only accept the RFC if we generally agree that PhantomData should be deprecated and replaced by more explicit annotations.

`PhantomData<T>` is not appropriate for dropck is because it always implements `Copy`. 

I think it could be appropriate if it were to only implement `Copy` if `T` is `Copy`. In this case for any `T` with drop glue we could require noop drop glue for `PhantomData` with the outlives requirements of `T`. However, such a `Copy` impl would be limiting other uses of `PhantomData` as it's quite useful to freely copy it around.

Because it implements `Copy` we're currently stuck with a type which has liveness requirements when involved in a drop, even though it doesn't get dropped itself. I personally label this behavior to clearly be a bug, hence the effort to change it via this RFC (and https://github.com/rust-lang/rust/issues/110288 for `[T; 0]`).

Reply:
So you're basically saying, PhantomData for dropck is 'even weirder' than for the other purposes. I can't disagree with that, for the specific `Copy` case you mention. However that's not the argument the RFC is making, so the RFC should be amended.

> (and https://github.com/rust-lang/rust/issues/110288 for [T; 0]).

`[T; 0]` [isn't copy](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=43f5d879c6dadc661fc4fc8e8a76024a) so that argument doesn't work. I'd rather special-case array length 0 in fewer cases than more. But anyway that's not part of this RFC.

Reply:
updated the motivation and mentioned auto traits + variance in the future possibilities section.